// dnd-core/dist/cjs/actions/dragDrop/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDragDropActions = void 0;
const beginDrag_1 = require("./beginDrag");
const publishDragSource_1 = require("./publishDragSource");
const hover_1 = require("./hover");
const drop_1 = require("./drop");
const endDrag_1 = require("./endDrag");
__exportStar(require("./types"), exports);
function createDragDropActions(manager) {
    return {
        beginDrag: (0, beginDrag_1.createBeginDrag)(manager),
        publishDragSource: (0, publishDragSource_1.createPublishDragSource)(manager),
        hover: (0, hover_1.createHover)(manager),
        drop: (0, drop_1.createDrop)(manager),
        endDrag: (0, endDrag_1.createEndDrag)(manager),
    };
}
exports.createDragDropActions = createDragDropActions;
