// dnd-core/dist/cjs/reducers/dragOperation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reduce = void 0;
const dragDrop_1 = require("../actions/dragDrop");
const registry_1 = require("../actions/registry");
const js_utils_1 = require("../utils/js_utils");
const initialState = {
    itemType: null,
    item: null,
    sourceId: null,
    targetIds: [],
    dropResult: null,
    didDrop: false,
    isSourcePublic: null,
};
function reduce(state = initialState, action) {
    const { payload } = action;
    switch (action.type) {
        case dragDrop_1.BEGIN_DRAG:
            return {
                ...state,
                itemType: payload.itemType,
                item: payload.item,
                sourceId: payload.sourceId,
                isSourcePublic: payload.isSourcePublic,
                dropResult: null,
                didDrop: false,
            };
        case dragDrop_1.PUBLISH_DRAG_SOURCE:
            return {
                ...state,
                isSourcePublic: true,
            };
        case dragDrop_1.HOVER:
            return {
                ...state,
                targetIds: payload.targetIds,
            };
        case registry_1.REMOVE_TARGET:
            if (state.targetIds.indexOf(payload.targetId) === -1) {
                return state;
            }
            return {
                ...state,
                targetIds: (0, js_utils_1.without)(state.targetIds, payload.targetId),
            };
        case dragDrop_1.DROP:
            return {
                ...state,
                dropResult: payload.dropResult,
                didDrop: true,
                targetIds: [],
            };
        case dragDrop_1.END_DRAG:
            return {
                ...state,
                itemType: null,
                item: null,
                sourceId: null,
                dropResult: null,
                didDrop: false,
                isSourcePublic: null,
                targetIds: [],
            };
        default:
            return state;
    }
}
exports.reduce = reduce;
