// dnd-core/dist/cjs/reducers/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reduce = void 0;
const dragOffset_1 = require("./dragOffset");
const dragOperation_1 = require("./dragOperation");
const refCount_1 = require("./refCount");
const dirtyHandlerIds_1 = require("./dirtyHandlerIds");
const stateId_1 = require("./stateId");
const js_utils_1 = require("../utils/js_utils");
function reduce(state = {}, action) {
    return {
        dirtyHandlerIds: (0, dirtyHandlerIds_1.reduce)(state.dirtyHandlerIds, {
            type: action.type,
            payload: {
                ...action.payload,
                prevTargetIds: (0, js_utils_1.get)(state, 'dragOperation.targetIds', []),
            },
        }),
        dragOffset: (0, dragOffset_1.reduce)(state.dragOffset, action),
        refCount: (0, refCount_1.reduce)(state.refCount, action),
        dragOperation: (0, dragOperation_1.reduce)(state.dragOperation, action),
        stateId: (0, stateId_1.reduce)(state.stateId),
    };
}
exports.reduce = reduce;
